packages/teleterm/src/ui/ClusterConnect/ClusterLogin/useClusterLogin.ts

```diff
@@ -43,36 +44,93 @@ export default function useClusterLogin(props: Props) {
     return authSettings;
   });
 
-  const [loginAttempt, login] = useAsync((opts: types.LoginParams) => {
-    refAbortCtrl.current = clustersService.client.createAbortController();
-    return clustersService.login(opts, refAbortCtrl.current.signal);
-  });
+  const [loginAttempt, login, setAttempt] = useAsync(
+    (authnType: PrimaryAuthType, opts: types.LoginParams) => {
+      refAbortCtrl.current = clustersService.client.createAbortController();
+      switch (authnType) {
+        case 'local':
+          return clustersService.loginLocal(
+            opts as types.LoginLocalParams,
+            refAbortCtrl.current.signal
+          );
+        case 'passwordless':
+          return clustersService.loginPasswordless(
+            opts as types.LoginPasswordlessParams,
+            refAbortCtrl.current.signal
+          );
+        default:
```

Comment:
I completely forgot to mention that to be fully "typesafe" it needs one more small thing.

https://github.com/gravitational/webapps/blob/331bdf1b34a28df80c7844ab51b58b51847c549e/packages/teleterm/src/ui/services/workspacesService/documentsService/documentsUtils.ts#L19-L26

That `(x: never): never` function makes it so that if someone adds a new type to `LoginParams`, the compiler will remind them that they need to handle this new case here.